qiime/core/type/signature.py

```diff
@@ -267,14 +282,20 @@ def function_to_signature(cls, function, inputs, parameters, outputs):
     # Artifacts come first followed by primitives.
 
     input_types = {}
-    for param_name, semantic_type in inputs.items():
-        view_type = function.__annotations__[param_name]
-        input_types[param_name] = (semantic_type, view_type)
-
     param_types = {}
-    for param_name, primitive_type in parameters.items():
-        view_type = function.__annotations__[param_name]
-        param_types[param_name] = (primitive_type, view_type)
+    defaults = {}
+    signature = inspect.signature(function)
+    for param_name, parameter in signature.parameters.items():
+        view_type = parameter.annotation
+        if param_name in inputs:
+            input_types[param_name] = inputs[param_name], view_type
+        elif param_name in parameters:
+            param_types[param_name] = parameters[param_name], view_type
+        elif param_name not in ('output_dir',):  # output_dir is special
```

Comment:
It'd be nice to not have to special-case here, since we don't know if this function is being called from `MethodSignature` or `VisualizerSignature` (`output_dir` only applies to visualizers). I think using the previous code (two loops) should work, something like:

```python
signature = inspect.signature(function)
input_types = {}
for param_name, semantic_type in inputs.items():
    parameter = signature.parameters[param_name]    

    view_type = parameter.annotation
    input_types[param_name] = (semantic_type, view_type)
    
    if parameter.default is not signature.empty:    
        defaults[param_name] = parameter.default

# Repeat this strategy, looping over `parameters`. Perhaps create a helper function since the code is the same.
```